src/modules/ttlmonitor/TTLMonitorInstrument.cpp

```diff
@@ -0,0 +1,159 @@
+// Copyright (C) 2022 European Spallation Source, ERIC. See LICENSE file
+//===----------------------------------------------------------------------===//
+///
+/// \file
+///
+/// \brief TTLMonitor is a dedicated module for TTL triggered beam monitor
+///
+//===----------------------------------------------------------------------===//
+
+#include <common/debug/Log.h>
+#include <common/debug/Trace.h>
+#include <common/time/TimeString.h>
+#include <ttlmonitor/TTLMonitorInstrument.h>
+#include <common/readout/vmm3/CalibFile.h>
+#include <common/readout/vmm3/Readout.h>
+#include <assert.h>
+
+// #undef TRC_LEVEL
+// #define TRC_LEVEL TRC_L_DEB
+
+namespace TTLMonitor {
+
+/// \brief load configuration and calibration files
+TTLMonitorInstrument::TTLMonitorInstrument(struct Counters & counters,
+  TTLMonitorSettings &moduleSettings,
+  EV42Serializer * serializer)
+    : counters(counters)
+    , ModuleSettings(moduleSettings)
+    , Serializer(serializer) {
+
+  XTRACE(INIT, ALW, "Loading configuration file %s",
+         ModuleSettings.ConfigFile.c_str());
+  Conf = Config(ModuleSettings.ConfigFile);
+  Conf.loadAndApply();
+
+  if (!ModuleSettings.FilePrefix.empty()) {
+    std::string DumpFileName = ModuleSettings.FilePrefix + "ttlmon_" + timeString();
+    XTRACE(INIT, ALW, "Creating HDF5 dumpfile: %s", DumpFileName.c_str());
+    DumpFile = VMM3::ReadoutFile::create(DumpFileName);
+  }
+
+  ESSReadoutParser.setMaxPulseTimeDiff(Conf.Parms.MaxPulseTimeDiffNS);
+}
+
+
+
+void TTLMonitorInstrument::processMonitorReadouts(void) {
+  ESSReadout::ESSTime & TimeRef = ESSReadoutParser.Packet.Time;
+  // All readouts are potentially now valid, negative TOF is not
+  // possible, but rings and fens
+  // could still be outside the configured range, also
+  // illegal time intervals can be detected here
+  assert(Serializer != nullptr);
+  Serializer->pulseTime(ESSReadoutParser.Packet.Time.TimeInNS); /// \todo sometimes PrevPulseTime maybe?
+
+  XTRACE(DATA, DEB, "processMonitorReadouts()");
+  for (const auto & readout : VMMParser.Result) {
+
+    if (DumpFile) {
+      dumpReadoutToFile(readout);
+    }
+
+    XTRACE(DATA, DEB, "readout: RingId %d, FENId %d, VMM %d, Channel %d, TimeLow %d",
+           readout.RingId, readout.FENId, readout.VMM, readout.Channel, readout.TimeLow);
+
+    if (readout.RingId/2 != 11) {
```

Comment:
Consider making ringid and fenid configurable

Reply:
done